### PR TITLE
Query cleanup

### PR DESCRIPTION
**Pull request by ADmad**

### CODE REVIEW

**Comment by ADmad** [comment]
The primary change done is the removed of self referencing closure in `Query::traverseExpressions()`.

Here are the results of test using https://github.com/ADmad/forked-php-orm-benchmark/tree/cakephp before and after modifying `Query::traverseExpressions()`

Before
```
|-------------------------------------------------------------------------------------------------------|
| Library                          | Insert | findPk | complex| hydrate|  with  | memory usage |  time  |
| --------------------------------:| ------:| ------:| ------:| ------:| ------:| ------------:| ------:|
|                              PDO |     21 |      4 |      9 |     73 |     96 |    1,671,816 |   0.21 |
|-------------------------------------------------------------------------------------------------------|
|                         AtlasOrm |    132 |     20 |    176 |    226 |    131 |    7,963,648 |   0.70 |
|-------------------------------------------------------------------------------------------------------|
|                         CycleOrm |    203 |     52 |    123 |    519 |    342 |   14,244,288 |   1.26 |
|-------------------------------------------------------------------------------------------------------|
|                        DoctrineM |     83 |     25 |    197 |    599 |    125 |   18,874,368 |   1.06 |
|               DoctrineMWithCache |     84 |     25 |    198 |    631 |    127 |   18,874,368 |   1.09 |
|            DoctrineMArrayHydrate |     85 |     26 |    227 |    283 |    124 |   16,777,216 |   0.78 |
|           DoctrineMScalarHydrate |    193 |     39 |    283 |    201 |     85 |   16,777,216 |   0.84 |
|          DoctrineMWithoutProxies |     86 |     26 |    220 |    442 |    212 |   18,874,368 |   1.02 |
|-------------------------------------------------------------------------------------------------------|
|                         Eloquent |    287 |     38 |    214 |    238 |    204 |    8,388,608 |   0.99 |
|             EloquentWithoutEvent |    258 |     36 |    213 |    186 |    192 |    6,291,456 |   0.90 |
|-------------------------------------------------------------------------------------------------------|
|                          CakeOrm |    526 |     71 |    408 |    427 |    301 |   16,343,888 |   1.74 |
|               CakeOrmNoHydration |    534 |     61 |    412 |    324 |    272 |   16,336,112 |   1.61 |
|-------------------------------------------------------------------------------------------------------|
```

After 
```
|-------------------------------------------------------------------------------------------------------|
| Library                          | Insert | findPk | complex| hydrate|  with  | memory usage |  time  |
| --------------------------------:| ------:| ------:| ------:| ------:| ------:| ------------:| ------:|
|                              PDO |     21 |      4 |      9 |     80 |    101 |    1,671,816 |   0.22 |
|-------------------------------------------------------------------------------------------------------|
|                         AtlasOrm |    135 |     21 |    183 |    232 |    118 |    7,963,648 |   0.70 |
|-------------------------------------------------------------------------------------------------------|
|                         CycleOrm |    204 |     53 |    122 |    479 |    324 |   14,244,288 |   1.20 |
|-------------------------------------------------------------------------------------------------------|
|                        DoctrineM |     83 |     25 |    201 |    612 |    126 |   18,874,368 |   1.07 |
|               DoctrineMWithCache |     84 |     25 |    195 |    594 |    120 |   18,874,368 |   1.05 |
|            DoctrineMArrayHydrate |     84 |     25 |    196 |    235 |     89 |   16,777,216 |   0.66 |
|           DoctrineMScalarHydrate |     82 |     25 |    191 |    171 |     81 |   16,777,216 |   0.58 |
|          DoctrineMWithoutProxies |     81 |     25 |    196 |    374 |    178 |   18,874,368 |   0.88 |
|-------------------------------------------------------------------------------------------------------|
|                         Eloquent |    285 |     38 |    206 |    237 |    198 |    8,388,608 |   0.97 |
|             EloquentWithoutEvent |    258 |     37 |    208 |    185 |    196 |    6,291,456 |   0.90 |
|-------------------------------------------------------------------------------------------------------|
|                          CakeOrm |    494 |     69 |    382 |    327 |    263 |    7,955,280 |   1.54 |
|               CakeOrmNoHydration |    486 |     56 |    383 |    249 |    245 |    7,947,504 |   1.43 |
|-------------------------------------------------------------------------------------------------------|
```

I am hoping someone that run the tests themselves using the benchmark I have used or their own to ensure the change does actually provide performance gain.

File: src/Database/Query.php [hunk 1]
```diff
@@ -1889,6 +1888,19 @@ public function decorateResults(?callable $callback, bool $overwrite = false)
         return $this;
     }
 
+    /**
+     * This function works similar to the traverse() function, with the difference
```
Comment (othercorey) [inline src/Database/Query.php]: Why did you duplicate this docblock?

File: src/Database/Query.php [hunk 2]
```diff
@@ -1903,25 +1915,46 @@ public function decorateResults(?callable $callback, bool $overwrite = false)
...
-                    $visitor($e);
-                }
+        if (!$callback instanceof Closure) {
+            $callback = Closure::fromCallable($callback);
+        }
 
-                return null;
-            }
+        $parts = array_keys($this->_parts);
+        foreach ($parts as $name) {
```
Comment (othercorey) [inline src/Database/Query.php]: Why not use `$parts as $name => $part` as above?

File: src/Database/Query.php [hunk 3]
```diff
@@ -1903,25 +1915,46 @@ public function decorateResults(?callable $callback, bool $overwrite = false)
...
+    /**
+     * Query parts traversal method used by traverseExpressions()
+     *
+     * @param \Cake\Database\ExpressionInterface|\Cake\Database\ExpressionInterface[] $expression Query expression or
+     *   array of expressions.
+     * @param \Closure $callback The callback to be executed for each ExpressionInterface
+     *   found inside this query.
+     * @return void
+     */
+    protected function _visitor($expression, Closure $callback)
```
Comment (othercorey) [inline src/Database/Query.php]: Can we call this `_partVisitor` or something more explicit?

Comment (othercorey) [inline src/Database/Query.php]: Missing `: void` return type.

**Comment by markstory** [comment]
Great work :clap:

**Pull request merged**